Merge feature/dashboard-oop to main

drewnoakes opened the pull request:
Fixes #1003
Fixes #851
Fixes #1753
Fixes #1688 
Fixes #1687 
Fixes #1686
Fixes #1639

This PR merges the feature branch for OOP dashboard work into `main`.

It will remain in draft until it is ready to merge.

The checklist of outstanding work (may be incomplete, but to be built up over the next day or two):

## Must have

- [x] Validate manifest generation, as that has broken when we've made similar changes in the past
- [x] Pass OLTP URL from user's AppHost project's environment (via `launchSettings.json`) to the dashboard process's environment when launching it
- [x] End to end testing
- [x] Validate that all claimed bug fixes are true
- [x] App model awaits dashboard readiness before starting other apps, so no telemetry is lost
- [x] Workload passes executable via dashboard attribute. Must either change to the DLL, or change the ApplicationExecutor to work with executables (not `dotnet <dll>`).

## Candidates

- [x] Remove default resource service URI from dashboard (see #1816)

![image](https://github.com/dotnet/aspire/assets/350947/82b1aae1-704f-4d37-b66f-2e01246e7998)

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/1774)

joperezr committed:
Add PR validation for feature branches.

drewnoakes committed:
Run dashboard out-of-process (#1667)

* Run dashboard out-of-process

This change causes the dashboard to be run as an orchestrated project, just like any other Aspire resource.

To achieve this, we:

- Break the project dependency from AppHost upon Dashboard.
  - Until we have support for dotnet tool resources, having the sample projects depend upon the dashboard directly, and register them with the DAB.
  - Copy `TimestampParser.DisplayFormat` into `Aspire.Hosting.Dashboard.ConsoleLogsConfigurationExtensions` where it was used, and add comments to keep them in sync.
- Have the dashboard wire up its own web application.
  - Remove `DashboardWebApplicationHost`.
  - Remove dashboard-specific code from `DistributedApplicationBuilder`.
  - Update URLs to resources defined in the dashboard. They no longer need (or work with) the `_content/Aspire.Dashboard/` prefix.
  - Remove workarounds/configuration to support static web assets, as the new structure supports this automatically.
  - Link `KnownProperties` and `KnownResourceTypes` into the hosting project, so both projects can share the definition without needing an assembly reference.
- Change the MSBuild SDK for the dashboard from Razor Class Library to Web project.
  - Remove explicit framework reference.
  - Remove redundant using directives (the new SDK has different implicit usings).
  - Make the dashboard an executable.
  - Update the CSS bundle file name (from `Aspire.Dashboard.bundle.scp.css` to `Aspire.Dashboard.styles.css`)

This work is incomplete for a few reasons:

1. Aspire users won't be able to create a project reference to the dashboard. That only works in this solution. We need support for resources defined via `dotnet tool` packages before we can fix that.
2. We don't launch a browser to the dashboard any more. You have to find the URL in the logs and launch it.

We will need to address those issues before the changes here can be merged.

---

To test this locally, run any of the sample applications. To find the dashboard's URL, navigate to `%TEMP%` and find the relevant log files. For example, `%TEMP%\aspire.oweozx1d.bmh\dashboard-5buv5so_out_eedbad51-0294-4ff8-9add-001f27499206` (your IDs will be different). The first log message should read _Now listening on: https://localhost:49337_ or similar. Open that URL in a browser.

* Make dashboard packable as a dotnet tool

Also adds a readme explaining dashboard configuration, and sets some metadata on the package for the icon, readme, etc.

mitchdenny committed:
Merge branch 'main' into feature/dashboard-oop

joperezr committed:
Merge branch 'main' into feature/dashboard-oop

mitchdenny committed:
Merge branch 'main' into feature/dashboard-oop

joperezr committed:
Merge pull request #1759 from dotnet-maestro-bot/merge/main-to-feature/dashboard-oop

[automated] Merge branch 'main' => 'feature/dashboard-oop'

danegsta committed:
[feature/dashboard-oop] Add initial workload targets and manifest entry for dashboard (#1705)

* Add initial workload targets and manifest entry for dashboard

* Do a better job handling trailing slashes for workload tool paths

* Update based on PR feedback

* Publishing as a tool by default breaks the workload build

* Remove extra parenthesis from import condition

* Updated workload package name

drewnoakes committed:
[dashboard-oop] Dynamic dashboard port (#1750)

* Expose resource service URIs on DashboardServiceHost

The resource service will no longer bind to a default, hardcoded port. This encourages port collisions. Instead, when no URL is set in the environment, a port is chosen dynamically.

When the app model launches the dashboard, it must pass the URIs exposed via `DashboardServiceHost.GetUrisAsync` via the `DOTNET_DASHBOARD_GRPC_ENDPOINT_URL` environment variable to the spawned instance. That work is happening concurrently, and will use this API is a subsequent change.

* Resource service listens on a single URI which must be loopback

* Require parsed URI to be absolute

danegsta committed:
[feature/dashboard-oop] Ensure the dashboard builds as a package for the workload (#1767)

* Ensure the dashboard actually builds as a package for the workload

* Publish then pack the dashboard as part of workload build

mitchdenny committed:
[dashboard-oop] Launch dashboard as exe (#1717)

* Run dashboard out-of-process

This change causes the dashboard to be run as an orchestrated project, just like any other Aspire resource.

To achieve this, we:

- Break the project dependency from AppHost upon Dashboard.
  - Until we have support for dotnet tool resources, having the sample projects depend upon the dashboard directly, and register them with the DAB.
  - Copy `TimestampParser.DisplayFormat` into `Aspire.Hosting.Dashboard.ConsoleLogsConfigurationExtensions` where it was used, and add comments to keep them in sync.
- Have the dashboard wire up its own web application.
  - Remove `DashboardWebApplicationHost`.
  - Remove dashboard-specific code from `DistributedApplicationBuilder`.
  - Update URLs to resources defined in the dashboard. They no longer need (or work with) the `_content/Aspire.Dashboard/` prefix.
  - Remove workarounds/configuration to support static web assets, as the new structure supports this automatically.
  - Link `KnownProperties` and `KnownResourceTypes` into the hosting project, so both projects can share the definition without needing an assembly reference.
- Change the MSBuild SDK for the dashboard from Razor Class Library to Web project.
  - Remove explicit framework reference.
  - Remove redundant using directives (the new SDK has different implicit usings).
  - Make the dashboard an executable.
  - Update the CSS bundle file name (from `Aspire.Dashboard.bundle.scp.css` to `Aspire.Dashboard.styles.css`)

This work is incomplete for a few reasons:

1. Aspire users won't be able to create a project reference to the dashboard. That only works in this solution. We need support for resources defined via `dotnet tool` packages before we can fix that.
2. We don't launch a browser to the dashboard any more. You have to find the URL in the logs and launch it.

We will need to address those issues before the changes here can be merged.

---

To test this locally, run any of the sample applications. To find the dashboard's URL, navigate to `%TEMP%` and find the relevant log files. For example, `%TEMP%\aspire.oweozx1d.bmh\dashboard-5buv5so_out_eedbad51-0294-4ff8-9add-001f27499206` (your IDs will be different). The first log message should read _Now listening on: https://localhost:49337_ or similar. Open that URL in a browser.

* Make dashboard packable as a dotnet tool

Also adds a readme explaining dashboard configuration, and sets some metadata on the package for the icon, readme, etc.

* Launching dashboard via DCP executable.

* Work in progress checkin.

* Remove unnecessary environment sets.

* Move dashboard path resolution to dcp options.

* PR feedback.

* Read ASPNETCORE_ENVIRONMENT ... from environment.

* Update src/Aspire.Hosting/Dashboard/DashboardServiceHost.cs

Co-authored-by: Drew Noakes <git@drewnoakes.com>

* Update src/Aspire.Hosting/Dashboard/DashboardServiceHost.cs

Co-authored-by: Drew Noakes <git@drewnoakes.com>

* Exclude dashboard resources from being watched.

* Filtering dashboard resources from view.

* Revert dashboard servicehost to parent branch state.

* Changes after integrating feature branch.

* Fix unit tests

---------

Co-authored-by: Drew Noakes <git@drewnoakes.com>

drewnoakes committed:
Rename and format (#1776)

* Formatting

Addresses diagnostics shown by the IDE. If we don't want these changes, we should reconfigure or disable those diagnostics.

* Rename method and log if delays are observed

* Replace `Trace.WriteLine` with `_logger.LogWarning`

drewnoakes committed:
Start dashboard first for playground scenario (#1775)

* Add KnownResourceNames

* Ensure dashboard is started first

When F5'ing one of the playground solutions in the dotnet/aspire repo, the dashboard is added as the last resource. This means it starts after all other resources, which is not what we want. This change ensures it is always started first.

mitchdenny committed:
Wait for dashboard to be online before starting other resources. (#1778)

* Add wait for dashboard start.

* WIP to merge with parent branch.

* Wait for dashboard if it is a Aspire resource.

* Add dashboard setup across playground.

* Update src/Aspire.Hosting/Dcp/ApplicationExecutor.cs

Co-authored-by: Drew Noakes <git@drewnoakes.com>

* Add short delay in polling loop.

* Fix up get request.

---------

Co-authored-by: Drew Noakes <git@drewnoakes.com>

drewnoakes committed:
Add API docs to DcpOptions (#1777)

danegsta committed:
[feature/dashboard-oop] Publish dashboard for specific RIDs before packing (#1771)

* Publish dashboard for specific RIDs before packing

* Add comments explaining the need to keep runtime lists in sync

* Remove unused item for icon

* Ensure it is clear the runtimes list matches the supported runtimes from DCP

* Allow parallel publishing

drewnoakes committed:
Fix build of OrleansAppHost (#1794)

drewnoakes committed:
Merge remote-tracking branch 'upstream/main' into feature/dashboard-oop

danegsta committed:
[dashboard-oop] Update dashboard attribute config for playground projects (#1799)

* Update attribute handling for playground projects

* Update comments on default path to dashboard

JamesNK commented on src/Aspire.Dashboard/Program.cs:
```diff
@@ -0,0 +1,7 @@
+
+using Aspire.Dashboard;
+
+// TODO potentially inline DashboardWebApplication in this file
+new DashboardWebApplication().Run();
```
I think how it now is fine.

JamesNK commented on src/Aspire.Dashboard/ConsoleLogs/TimestampParser.cs:
```diff
@@ -5,9 +5,11 @@
 {
-    public const string DisplayFormat = "yyyy-MM-ddTHH:mm:ss.fffffff";
+    // NOTE: Must match Aspire.Hosting.Dashboard.ConsoleLogsConfigurationExtensions.DisplayFormat
+    private const string DisplayFormat = "yyyy-MM-ddTHH:mm:ss.fffffff";
```
Move to a shared const file?

JamesNK commented on src/Aspire.Hosting/Dashboard/ConsoleLogsConfigurationExtensions.cs:
```diff
@@ -1,13 +1,15 @@
 internal static class ConsoleLogsConfigurationExtensions
 {
+    // NOTE: Must match Aspire.Dashboard.ConsoleLogs.TimestampParser.DisplayFormat
+    private const string DisplayFormat = "yyyy-MM-ddTHH:mm:ss.fffffff";
```
Move to a const file?

JamesNK commented on src/Aspire.Hosting/Dashboard/DashboardServiceHost.cs:
```diff
@@ -1,113 +1,188 @@
         KubernetesService kubernetesService,
         IOptions<PublishingOptions> publishingOptions,
         ILoggerFactory loggerFactory,
+        ILogger<DashboardServiceHost> logger,
```
Since ILoggerFactory is already being passed in then the logger should be created from the factory inside the constructor.

drewnoakes replied:
Why would that make a difference?

JamesNK replied:
Passing in both is unusual. At runtime the DI container creates the extra logger for you. The only difference it would make would be when writing unit tests. There would be one less argument to pass in.

JamesNK commented on src/Aspire.Hosting/Dashboard/DashboardServiceHost.cs:
```diff
@@ -1,113 +1,188 @@
+    /// </remarks>
+    public async Task<string> GetResourceServiceUriAsync(CancellationToken cancellationToken = default)
+    {
+        var stopwatch = Stopwatch.StartNew();
```
nit: Use ValueStopwatch (it's in the shared source)

JamesNK commented on src/Aspire.Hosting/Dashboard/DashboardServiceHost.cs:
```diff
@@ -1,113 +1,188 @@
+            publishingOptions.Value.Publisher == "manifest") // HACK: Manifest publisher check is temporary until DcpHostService is integrated with DcpPublisher.
         {
-            // HACK: Manifest publisher check is temporary until DcpHostService is integrated with DcpPublisher.
+            _resourceServiceUri.SetCanceled();
```
Write log that the dashboard service isn't starting?

JamesNK replied:
I wanted to see what would happen if the dashboard is disabled but DistributedApplicationOptions's DashboardEnabled/DisableDashboard (it's weird there is two) properties are never set.

Is it for a future feature?

drewnoakes replied:
Added a log message.

Will kick off a conversation about disabling. Supposedly there's a mechanism for this, though unless there's some reflection at play, the only place that sets `DisableDashboard` (from which `DashboardEnabled` is derived) is in test code.

JamesNK commented on src/Aspire.Hosting/Dcp/ApplicationExecutor.cs:
```diff
@@ -85,7 +107,100 @@ public async Task RunApplicationAsync(CancellationToken cancellationToken = defa
+            WorkingDirectory = dashboardWorkingDirectory
+        };
+
+        var grpcEndpointUrl = await _dashboardServiceHost.GetResourceServiceUriAsync(cancellationToken).ConfigureAwait(false);
```
`GetResourceServiceUriAsync` could throw an error. Is error handling needed here? e.g.
```suggestion
        try
        {
            var grpcEndpointUrl = await _dashboardServiceHost.GetResourceServiceUriAsync(cancellationToken).ConfigureAwait(false);
        }
        catch (Exception ex)
        {
            throw new DistributedApplicationException("Error when getting resource service URL.", ex);
        }
```

JamesNK commented on src/Aspire.Hosting/Dcp/ApplicationExecutor.cs:
```diff
@@ -85,7 +107,100 @@ public async Task RunApplicationAsync(CancellationToken cancellationToken = defa
+                _logger.LogDebug(ex, "Dashboard not ready yet.");
+            }
+
+            await Task.Delay(TimeSpan.FromMilliseconds(50), cancellationToken).ConfigureAwait(false);
```
Unless the CT is raised an precisisly the right moment (after `Task.Delay` is finished but before `cancellationToken.IsCancellationRequest` is check) then on timeout this method will throw `OperationCanceledException` from `Task.Delay`. The friendly `DistributedApplicationException` below won't be thrown.

Need to wrap loop in a try catch.

JamesNK commented on src/Aspire.Hosting/Dcp/ApplicationExecutor.cs:
```diff
@@ -85,7 +107,100 @@ public async Task RunApplicationAsync(CancellationToken cancellationToken = defa
+            await Task.Delay(TimeSpan.FromMilliseconds(50), cancellationToken).ConfigureAwait(false);
+        }
 
+        throw new DistributedApplicationException("Timed out waiting for dashboard to be responsive.");
```
Include how long the timeout is in the error message. I think users would appreciate knowing how long the timeout takes to trigger.

drewnoakes committed:
[dashboard-oop] Support starting dashboard as DLL or executable (#1798)

* Support starting dashboard as DLL or executable

The workload ships an executable, but `ApplicationExecutor` was written to run a DLL via `dotnet <dll>`.

This change checks whether the dashboard is a DLL or not, and configures the executable spec accordingly.

* Fix compile error

Co-authored-by: David Negstad <50252651+danegsta@users.noreply.github.com>

---------

Co-authored-by: David Negstad <50252651+danegsta@users.noreply.github.com>

JamesNK commented on src/Aspire.Hosting/Dcp/ApplicationExecutor.cs:
```diff
@@ -85,7 +107,110 @@ public async Task RunApplicationAsync(CancellationToken cancellationToken = defa
+            WorkingDirectory = dashboardWorkingDirectory
+        };
+
+        if (StringComparer.OrdinalIgnoreCase.Equals(".dll", Path.GetExtension(fullyQualifiedDashboardPath)))
```
string.Equals + StringComparison.OridinalIgnoreCase

drewnoakes replied:
Why do you prefer that? I benchmarked it once and it's marginally faster, but not enough to make a difference here. This is slightly less verbose.

JamesNK reviewed (commented):
The PR is a draft and things are still in progress, but since time is important, I've added some comments.

Edit: I'll just make a PR to the branch myself to fix these.

mitchdenny committed:
Fix dashboard startup ordering. (#1804)

* Fix dashboard startup ordering.

* Workaround pathing bug.

* Fix up AspireDashboardDir (keep it in Directory.Build.props in repo root).

mitchdenny commented on playground/eShopLite/AppHost/Properties/launchSettings.json:
```diff
@@ -8,7 +8,9 @@
-        "DOTNET_DASHBOARD_OTLP_ENDPOINT_URL": "http://localhost:16031"
+        "DOTNET_DASHBOARD_OTLP_ENDPOINT_URL": "http://localhost:16031",
+        "DOTNET_DASHBOARD_GRPC_ENDPOINT_URL": "http://localhost:5555"
+        //        "DOTNET_ASPIRE_SHOW_DASHBOARD_RESOURCES": "true"
```
Intentional. So folks who are working with the dashboard can easily uncomment it and use it to look at the dashboard telemtry etc.

JamesNK commented:
Now that the dashboard runs in a separate process, how do we view console logs for the dashboard in local development?

mitchdenny commented:
> Now that the dashboard runs in a separate process, how do we view console logs for the dashboard in local development?

Uncomment this line:
https://github.com/dotnet/aspire/blob/fa67232ec8fc3c051db17aa91760dd7199e35087/playground/eShopLite/AppHost/Properties/launchSettings.json#L13

Ideally you will also have the dashboard added as a resource (the default in the source) so that the debugger will get attached as well.

JamesNK commented:
Ok. Shouldn't it always be uncommented for our sandbox projects?

mitchdenny commented:
> Ok. Shouldn't it always be uncommented for our sandbox projects?

Yeah that is fair. Did you want to put it in your catch all PR?

JamesNK commented:
I'll do it in my catch all 👻

JamesNK commented:
I now see the dashboard in the dashboard 🔥 Good: Console logs are available. Bad: Other telemetry (structured logs, tracing, metrics) isn't.

We don't need it for P3, but it would be great if the dashboard sent telemetry about the dashboard to the dashboard.

drewnoakes commented:
The test `Aspire.Azure.Storage.Queues.Tests.ConformanceTests.TracingEnablesTheRightActivitySource` fails intermittently but fairly regularly.

drewnoakes commented:
/azp run

JamesNK committed:
[dashboard-oop] PR feedback (#1805)

JamesNK commented on src/Aspire.Dashboard/README.md:
```diff
@@ -0,0 +1,11 @@
+
+- `DOTNET_DASHBOARD_OTLP_ENDPOINT_URL` specifies the OTLP endpoint, and defaults to http://localhost:18889.
+- `ASPNETCORE_URLS` specifies the HTTP endpoint via which the dashboard web application is served, and defaults to http://localhost:18888.
+- `DOTNET_DASHBOARD_GRPC_ENDPOINT_URL` specifies the gRPC endpoint to which the dashboard connects to for its data, and defaults to http://localhost:18999.
```
Is `DOTNET_DASHBOARD_GRPC_ENDPOINT_URL` the right name? When I read it, I think that it is a grpc endpoint on the dashboard. For example, `DOTNET_DASHBOARD_OTLP_ENDPOINT_URL` is the dashboard's OTLP endpoint address.

One of these isn't right.

drewnoakes replied:
Good point. Perhaps `DOTNET_RESOURCE_SERVICE_ENDPOINT_URL` is a better name.

drewnoakes replied:
It also no longer has a default. I'll update that too.

drewnoakes replied:
> It also no longer has a default. I'll update that too.

Actually, the default is removed in https://github.com/dotnet/aspire/pull/1795, so I'll update the doc in that PR instead.

JamesNK commented on src/Aspire.Hosting/Dashboard/DashboardExtensions.cs:
```diff
@@ -0,0 +1,21 @@
+/// </summary>
+public static class DashboardExtensions
+{
+    public static IResourceBuilder<ExecutableResource> AddDashboard(this IDistributedApplicationBuilder builder)
```
Should this be in P3?

drewnoakes replied:
Removed. It was WIP code that stuck around. Thanks.

drewnoakes committed:
Rename environment variable

This hasn't shipped anywhere yet, so it's fine to rename it now.

drewnoakes committed:
Remove unused DashboardExtensions

drewnoakes commented on src/Aspire.Hosting/Dashboard/DashboardServiceHost.cs:
```diff
@@ -38,76 +54,134 @@ internal sealed class DashboardServiceHost : IHostedService
+
+            if (addressFeature is null)
+            {
+                _resourceServiceUri.SetException(new InvalidOperationException("Could not obtain IServerAddressesFeature. Dashboard URIs are not available."));
```
We modelled multiple values here previously, then changed to singular. This needs updating:

```suggestion
                _resourceServiceUri.SetException(new InvalidOperationException("Could not obtain IServerAddressesFeature. Resource service URI is not available."));
```

drewnoakes committed:
Update exception message

drewnoakes committed:
DashboardServiceHost logs when skipping hosting resource service

drewnoakes commented:
There are two open PRs currently targeting this branch, which will be retargeted to `main`:

- https://github.com/dotnet/aspire/pull/1795
- https://github.com/dotnet/aspire/pull/1796

drewnoakes commented:
I reviewed the issues this PR claims to fix and am happy that they're all addressed by this change:

- #1003
- #851
- #1753
- #1688 
- #1687 
- #1686
- #1639

drewnoakes commented on src/Aspire.Dashboard/README.md:
```diff
@@ -0,0 +1,11 @@
+# .NET Aspire Dashboard
+
+## As a dotnet tool
+
+The dashboard can be packaged and executed as a `dotnet tool`.
```
Apologies, the docs weren't updated when the tool support was removed.

We removed it because we no longer needed it for our scenario and (from memory) it conflicted with other packaging requirements.

Do you have a use case for a tool? If so, could you file an issue explaining how you'd use it?

drewnoakes replied:
I think you should be able to get that executable working and not need a dotnet tool. If not, please open a new issue.
